src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs

```diff
@@ -377,63 +379,115 @@ public static unsafe ConnectionOptions ParseConnection(StringValues connection)
             return connectionOptions;
         }
 
-        public static unsafe TransferCoding GetFinalTransferCoding(StringValues transferEncoding)
+        public static TransferCoding GetFinalTransferCoding(StringValues transferEncoding)
         {
+            const char lowerCaseChar = (char)0x0020;
+            const ulong lowerCaseULong = 0x0020_0020_0020_0020;
+            const uint lowerCaseUInt = 0x0020_0020;
+
+            const ulong hunkChars = 0x006b_006e_0075_0068; // 4 chars "hunk"
+            const uint edChars = 0x0064_0065; // 2 chars "ed"
+
             var transferEncodingOptions = TransferCoding.None;
 
             var transferEncodingCount = transferEncoding.Count;
             for (var i = 0; i < transferEncodingCount; i++)
             {
-                var value = transferEncoding[i];
-                fixed (char* ptr = value)
-                {
-                    var ch = ptr;
-                    var tokenEnd = ch;
-                    var end = ch + value.Length;
+                var value = transferEncoding[i].AsSpan();
 
-                    while (ch < end)
+                while (value.Length > 0)
+                {
+                    int offset;
+                    char c = '\0';
+                    // Skip any spaces and empty values.
+                    for (offset = 0; offset < value.Length; offset++)
                     {
-                        while (tokenEnd < end && *tokenEnd != ',')
+                        c = value[offset];
+                        if (c == ' ' || c == ',')
                         {
-                            tokenEnd++;
+                            continue;
                         }
 
-                        while (ch < tokenEnd && *ch == ' ')
+                        break;
+                    }
+
+                    // Skip last read char.
+                    offset++;
+                    if ((uint)offset > (uint)value.Length)
+                    {
+                        // Consumed entire string, move to next.
+                        break;
+                    }
+
+                    // Remove leading spaces or empty values.
+                    value = value.Slice(offset);
+                    offset = 0;
+
+                    var byteValue = MemoryMarshal.AsBytes(value);
+
+                    if ((c | lowerCaseChar) == 'c' &&
```

Comment:
I see this (good) pattern repeated several times (also in the other PRs).

Maybe this can be refactored with the [struct-trick](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAIBlAC2ygAcAZbYAOgCUBXAOwwEt8MANwBYAFDi+vGFG7YANuSkYZAM2xgY5AJIB5evwjdxAb3HkL5TsvIA1BZy0nyAcxgYh5AL7ifEscoycorEAEzkAMLkZmKW5ADaAFJ8GADiMNwyfGAAFBgAnvQwEKo5EdQ8ANb6htwAlHUAuuaWxADM5MAQEIoREPj0dDAAPAAqNXxGAHw51rzkAG4OMGjk4waT3OQQG0Z15ADu1DJa67XkIDoTRi0WMXFxxADsi8vkALzv27vcrPbyjlEsUsfjit3I9CgfCWKnIuAwUE4YAwkQq3GqPwuVx+4PuDws7SsNn+jg+U3IpAQpFIoVIAH1qbSgXE/F4gA===) (I'm sure you get the idea)

But I don't know if this is a win code-wise. The constants (see comment above) could be wrapped in more obvious names that way and the logic re-used.
Perf-wise it should result in the same.